Update benchmark suite for poliastro v0.14

samtx opened the pull request:
This updates the existing benchmark suite to work with the new poliastro v0.14 APIs and uses flit for build and install.

This is still a work in progress.

samtx committed:
Update Poliastro benchmark suite for poliastro 0.14

- Begin by updating the asv.conf.json
- There is still a problem with conda and asv, so this is a work in progress

samtx committed:
WIP: Trying to get the benchmarks to pass

- Only use python 3.8 for now
- Remove all but one benchmark tests
- Add some dummy benchmarks that should pass... but they don't
- The benchmarks are still failing

astrojuanlu commented on asv.conf.json:
```diff
@@ -10,10 +12,43 @@
-        "numba": ["0.39"]
+        "numpy": [],
+        "numba": [],
+        "astropy": [],
```
Do we need to specify all the dependencies here?

astrojuanlu commented on asv.conf.json:
```diff
@@ -10,10 +12,43 @@
     "matrix": {
-        "numpy": ["1.14"],
-        "numba": ["0.39"]
+        "numpy": [],
```
Perhaps we should update the versions to something more recent, for example NumPy 1.16 and numba 0.46, but keep the versions anyway. This way we can control for performance changes when the dependencies upgrade.

astrojuanlu reviewed (commented):
Left a couple of comments, although I see this is still a work in progress. Thanks @samtx for this pull request! It's really needed.

samtx committed:
Now the imports are working correctly
- Remove extraneous benchmarks for testing

samtx committed:
Move original benchmarks back to benchmarks folder
- Use virtualenv instead of conda
- Still some issues with APIs in perturbations.py

astrojuanlu commented on benchmarks/perturbations.py:
```diff
@@ -67,5 +69,6 @@ def setup(self):
-        cowell(self.orbit, self.tof, ad=atmospheric_drag, R=self.R, C_D=self.C_D, A=self.A, m=self.m, H0=self.H0, rho0=self.rho0)
+        
+    def time_atmospheric_drag_exponential(self):
+        cowell(self.orbit, self.tof, ad=atmospheric_drag_exponential, R=self.R, C_D=self.C_D, A=self.A, m=self.m, H0=self.H0, rho0=self.rho0)
```
Oops, this API changed as well. This should be:

```suggestion
        cowell(self.orbit.attractor.k, self.orbit.r, self.orbit.v, self.tof.reshape(-1), ad=atmospheric_drag_exponential, R=self.R, C_D=self.C_D, A=self.A, m=self.m, H0=self.H0, rho0=self.rho0)
```

and the same for all `cowell` calls.

astrojuanlu commented on benchmarks/perturbations.py:
```diff
@@ -67,5 +69,6 @@ def setup(self):
         self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
```
```suggestion
        self.rho0 = rho0_earth.to_value(u.kg / u.km ** 3)
```

and this comes from `from poliastro.constants import rho0_earth`

samtx committed:
update Earth density APIs

samtx commented:
There are still a few errors:

```
[ 80.00%]
··· perturbations.AtmosphericDrag.time_atmospheric_drag_exponential                                     failed[ 80.00%] ···· Traceback (most recent call last):
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1052, in main_run
                   skip = benchmark.do_setup()
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 583, in do_setup
                   result = Benchmark.do_setup(self)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 515, in do_setup
                   setup(*self._current_params)
                 File "/home/sam/poliastro_benchmarks/benchmarks/perturbations.py", line 71, in setup
                   self.rho0 = rho0_earth.to(u.kg / u.km ** 3)  # kg/km^3
               AttributeError: 'SolarSystemPlanet' object has no attribute 'H0'

[ 90.00%] ··· perturbations.J2_Propagation.time_J2_propagation                                                    failed[ 90.00%] ···· Traceback (most recent call last):
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1058, in main_run
                   result = benchmark.do_run()
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 537, in do_run
                   return self.run(*self._current_params)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 627, in run
                   samples, number = self.benchmark_timing(timer, min_repeat, max_repeat,
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 663, in benchmark_timing
                   timing = timer.timeit(number)
                 File "/usr/lib/python3.8/timeit.py", line 176, in timeit
                   timing = self.inner(it, self.timer)
                 File "<timeit-src>", line 6, in inner
                 File "/home/sam/poliastro_benchmarks/benchmarks/perturbations.py", line 31, in time_J2_propagation
                   def time_J2_propagation(self):
               TypeError: cowell() missing 2 required positional arguments: 'v' and 'tofs'

[100.00%] ··· perturbations.J3_Propagation.time_J3_propagation                                                    failed[100.00%] ···· Traceback (most recent call last):
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1184, in main_run_server
                   main_run(run_args)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 1058, in main_run
                   result = benchmark.do_run()
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 537, in do_run
                   return self.run(*self._current_params)
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 627, in run
                   samples, number = self.benchmark_timing(timer, min_repeat, max_repeat,
                 File "/home/sam/poliastro_benchmarks/venv/lib/python3.8/site-packages/asv/benchmark.py", line 663, in benchmark_timing
                   timing = timer.timeit(number)
                 File "/usr/lib/python3.8/timeit.py", line 176, in timeit
                   timing = self.inner(it, self.timer)
                 File "<timeit-src>", line 6, in inner
                 File "/home/sam/poliastro_benchmarks/benchmarks/perturbations.py", line 52, in time_J3_propagation
                   def time_J3_propagation(self):
               TypeError: cowell() missing 2 required positional arguments: 'v' and 'tofs'
```

astrojuanlu commented on benchmarks/perturbations.py:
```diff
@@ -64,8 +68,9 @@ def setup(self):
         # parameters of the atmosphere
-        self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
+        self.rho0 = rho0_earth.to(u.kg / u.km ** 3)  # kg/km^3
         self.H0 = Earth.H0.to(u.km).value
```
This is now `from poliastro.constants import H0_earth`

astrojuanlu commented on benchmarks/perturbations.py:
```diff
@@ -64,8 +68,9 @@ def setup(self):
 
         # parameters of the atmosphere
-        self.rho0 = Earth.rho0.to(u.kg / u.km**3).value  # kg/km^3
+        self.rho0 = rho0_earth.to(u.kg / u.km ** 3)  # kg/km^3
```
Don't we need to use the `.value` here? Otherwise it's an Astropy quantity

astrojuanlu reviewed (commented):
Left two more comments, we're getting closer!

samtx committed:
Fix pertubations.py API. It's alive!

astrojuanlu commented on asv.conf.json:
```diff
@@ -1,32 +1,65 @@
+        "matplotlib": "",
     },
 
+    "include": [
```
xref: https://asv.readthedocs.io/en/stable/asv.conf.json.html#include

samtx replied:
This is to make asv create only two environments for benchmarking:
- py3.8 with dependencies
- py3.7 with dependencies and cached_property

samtx committed:
Remove the 'pip+' from the cached_property dependency since we are using virtualenv now

astrojuanlu commented on asv.conf.json:
```diff
@@ -1,32 +1,65 @@
     "matrix": {
-        "numpy": ["1.14"],
-        "numba": ["0.39"]
+        "numpy": "",
+        "numba": "",
```
What about fixing the versions? That way we control when do we upgrade the benchmarks and notice when a new dependency introduces a performance change. I propose:

```suggestion
        "numpy": ["1.17.5"],
        "numba": ["0.47.0"],
        "llvmlite": ["0.31.0"]
```

these versions are the oldest ones with Python 3.8 wheels, I think.

astrojuanlu commented on asv.conf.json:
```diff
@@ -1,32 +1,65 @@
-        "numba": ["0.39"]
+        "numpy": "",
+        "numba": "",
+        "astropy": "",
```
Do we need to explicitly include all the extra dependencies? If we will just use the newest one, we should not need it, am I right?

astrojuanlu commented on asv.conf.json:
```diff
@@ -1,32 +1,65 @@
 
-    "environment_type": "conda",
-    "conda_channels": ["conda-forge", "defaults"],
+    // "environment_type": "virtualenv",
+    "environment_type": "virtualenv",
```
Remove duplicated line:

```suggestion
    "environment_type": "virtualenv",
```

astrojuanlu commented on asv.conf.json:
```diff
@@ -1,32 +1,65 @@
+
     // The versions of Python to test against
-    "pythons": ["3.6"],
+    "pythons": ["3.8"], // Python 3.7 included below with extra cached_property dependency
```
:+1:

astrojuanlu commented on benchmarks/kepler_propagation.py:
```diff
@@ -1,17 +1,37 @@
+    for param in suite.params:
+        suite.setup()
+        suite.time_propagation(method=method, ecc=ecc)
+    
```
Remove final line without newline character?

astrojuanlu commented on benchmarks/perturbations.py:
```diff
@@ -43,29 +45,56 @@ def setup(self):
+
+    suite = AtmosphericDrag()
+    suite.setup()
+    suite.time_atmospheric_drag_exponential()
```
Add newline at the end of the file?

astrojuanlu reviewed (changes requested):
Left a few style comments, other than that this is good to go. Also, do you think we can move the files back to a `benchmarks/` directory?

samtx committed:
Use Black formatting

samtx committed:
Add lambert.py back to benchmarks
- Fix formatting of asv.conf.json
- Format lambert.py with black

samtx committed:
Change suite name from Propagation -> KeplerPropagation

samtx commented:
Some more fixes:
- I used `black` to format everything.
- I added `lambert.py` back to the benchmark suite. I had missed it earlier.
- I tried fixing the dependency version of numpy and numba, but kept getting compilation errors when asv tried to install numpy into the virtual environments. Keeping the dependencies as the latest version was the only option that did not cause any issues. There could be a way to fix the dependency version, but I think that it should be the oldest versions required in poliastro's `pyproject.toml`. I don't think asv automatically re-downloads new dependency versions because then the whole benchmark suite would need to be rerun for all applicable commits. So the dependency versions should be fixed to the latest version available at the time the environment is created.
- I've only tested the benchmark suite with python 3.8. I'm currently installing python 3.7 to confirm that it works as well. 

Another thing left to do would be to make the `README.rst` more thorough. The benchmarks themselves look good.

Also, the reason that the parameterized functions in `Lambert` and `KeplerPropagation` are nested in dictionaries is so that the names appear nicely in the charts. If you parameterize using the function handle directly then you get the `repr()` value of the function in the chart legend.

astrojuanlu commented:
Thanks a lot @samtx! I think the improvements in the README and the question about the versions can be left for future pull requests. Let's merge this already, and strive for incremental improvements. I will actually try to run the benchmarks again with a wide variety of commits in a dedicated machine.
